docs/fomega/cek-cps-experiments/src/CekMachine.recursive.hs

```diff
@@ -0,0 +1,138 @@
+
+-- | Look up a name in an environment.
+lookupName :: Name () -> Environment -> Maybe Closure
+lookupName name (Environment env) = IntMap.lookup (unUnique $ nameUnique name) env
+
+
+evalCek :: Closure -> Closure
+evalCek !cl@(Closure term env) =
+--    trace ("eval: " ++ termStr term) $
+    case term of
+      TyAbs{} -> cl
+      LamAbs{} -> cl
+      Constant{} -> cl
+
+      TyInst _ fun ty -> instantiateEvaluate ty $ evalCek (Closure fun env)
+
+      Apply _ fun arg -> applyEvaluate (evalCek $ Closure fun env) (evalCek $ Closure arg env)
+
+      Wrap ann tyname ty term' ->
+          let Closure term2 env2 = evalCek $ Closure term' env
+          in  Closure (Wrap ann tyname ty term2) env2
+
+      Unwrap _ term' ->
+             case evalCek $ Closure term' env of
+               Closure (Wrap _ _ _ t) env' -> evalCek $ Closure t env'
+               _                           -> throw $ MachineException NonWrapUnwrappedMachineError term'
+
+      Error{} -> throw $ MachineException OpenTermEvaluatedMachineError term
+
+      Var _ name -> case lookupName name env of
+                      Nothing                     -> throw $ MachineException OpenTermEvaluatedMachineError term
+                      Just cl@(Closure val7 env7) -> cl
+
+-- | Instantiate a term with a type and proceed.
+-- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
+-- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
+-- apply the term to the type via 'TyInst'.
+instantiateEvaluate
+    :: Type TyName () -> Closure -> Closure
+instantiateEvaluate _  !(Closure (TyAbs _ _ _ body) env) = evalCek (Closure body env)
+instantiateEvaluate ty !(Closure fun env)
+    | isJust $! termAsPrimIterApp fun = Closure (TyInst () fun ty) env
+    | otherwise                      = throw $ MachineException NonPrimitiveInstantiationMachineError fun
+
+-- | Apply a function to an argument and proceed.
+-- If the function is a 'LamAbs', then extend the current environment with a new variable and proceed.
+-- If the function is not a 'LamAbs', then 'Apply' it to the argument and view this
+-- as an iterated application of a 'BuiltinName' to a list of 'Value's.
+-- If succesful, proceed with either this same term or with the result of the computation
+-- depending on whether 'BuiltinName' is saturated or not.
+applyEvaluate :: Closure -> Closure -> Closure
+applyEvaluate !(Closure funVal funEnv) !(Closure argVal argEnv) =
+    case funVal of
+      LamAbs _ name _ body -> evalCek $ Closure body (extendEnvironment name argVal argEnv funEnv)
+
+      _ -> let !term = Apply () funVal argVal in
+           case termAsPrimIterApp term of
+            Nothing                       ->
+                throw $ MachineException NonPrimitiveApplicationMachineError term
+            Just (IterApp headName spine) ->
+                case runQuote $! ((applyBuiltinName $! headName) $! spine) of
```

Comment:
In `runQuote $!` the `$!` has no effect, because this thing is inside pattern matching `case`. Forcing `headName` does not have any effect, because it's forced in `applyBuiltinName` anyway. Forcing `spine` does not do much, because you anyway only force to WHNF and the spine is forced in `applyBuiltinName` as well.